sycl/include/CL/sycl/detail/group_sort_impl.hpp

```diff
@@ -0,0 +1,224 @@
+      const std::size_t shift_1 = idx - start_1;
+      const std::size_t shift_2 = l_search_bound_2 - start_2;
+
+      out_acc1[start_out + shift_1 + shift_2] = intermediate_item_1;
+    }
+  }
+  // Process 2nd sequence
+  if (local_start_2 < local_end_2) {
+    // Reduce the range for searching within the 1st sequence and handle bound
+    // items find left border in 1st sequence
+    const auto local_l_item_2 = in_acc1[local_start_2];
+    std::size_t l_search_bound_1 =
+        my_upper_bound(in_acc1, start_1, end_1, local_l_item_2, comp);
+    const std::size_t l_shift_1 = l_search_bound_1 - start_1;
+    const std::size_t l_shift_2 = local_start_2 - start_2;
+
+    out_acc1[start_out + l_shift_1 + l_shift_2] = local_l_item_2;
+
+    std::size_t r_search_bound_1{};
+    // find right border in 1st sequence
+    if (local_size_2 > 1) {
+      const auto local_r_item_2 = in_acc1[local_end_2 - 1];
+      r_search_bound_1 = my_upper_bound(in_acc1, l_search_bound_1, end_1,
+                                        local_r_item_2, comp);
+      const std::size_t r_shift_1 = r_search_bound_1 - start_1;
+      const std::size_t r_shift_2 = local_end_2 - 1 - start_2;
+
+      out_acc1[start_out + r_shift_1 + r_shift_2] = local_r_item_2;
+    }
+
+    // Handle intermediate items
+    for (auto idx = local_start_2 + 1; idx < local_end_2 - 1; ++idx) {
+      const auto intermediate_item_2 = in_acc1[idx];
+      // we shouldn't seek in whole 1st sequence. Just for the part where the
+      // 2nd sequence should be
+      l_search_bound_1 =
+          my_upper_bound(in_acc1, l_search_bound_1, r_search_bound_1,
+                         intermediate_item_2, comp);
+      const std::size_t shift_1 = l_search_bound_1 - start_1;
+      const std::size_t shift_2 = idx - start_2;
+
+      out_acc1[start_out + shift_1 + shift_2] = intermediate_item_2;
+    }
+  }
+}
+
+template <typename Iter, typename Compare>
+void bubble_sort(Iter first, const std::size_t begin, const std::size_t end,
+                 Compare comp) {
+  if (begin < end) {
+    for (std::size_t i = begin; i < end; ++i) {
+      // Handle intermediate items
+      for (std::size_t idx = i + 1; idx < end; ++idx) {
+        if (comp(first[idx], first[i])) {
+          my_swap(first[i], first[idx]);
+        }
+      }
+    }
+  }
+}
+
+template <typename Group, typename Iter, typename Compare>
+void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
+                std::uint8_t *scratch) {
+  using T = typename std::iterator_traits<Iter>::value_type;
+  auto id = cl::sycl::detail::Builder::getNDItem<Group::dimensions>();
+  const std::size_t idx = id.get_local_id();
+  const std::size_t local = group.get_local_range().size();
+  const std::size_t chunk = (n - 1) / local + 1;
+
+  // we need to sort within work item firstly
+  bubble_sort(first, idx * chunk, sycl::min((idx + 1) * chunk, n), comp);
+  id.barrier();
+
+  T *temp = reinterpret_cast<T *>(scratch);
```

Comment:
Is this always guaranteed to be safe?  Should we replace `std::uint8_t` above with `char` or `byte`?

Reply:
While it is exceptionally hard in 2021 to come across the system, that thinks that byte is anything but 8 bits, I agree with John, that the standard only guarantees this to be safe for `unsigned char`, `char`, and `byte`. And yet, I would still prefer this last parameter to be replaced with `sycl::span` or something like that to ensure that we do not get out of bounds of this array. (The last statement probably requires some spec work).

Reply:
I'll think about it. Thanks for the comment

Reply:
> Should we replace std::uint8_t above with char or byte?

If so, it's better to make changes for the Spec (https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc).
What do you think? What is a better way to reflect it in the Spec?

Reply:
>the standard only guarantees this to be safe for unsigned char, char, and byte

Could you please point on the C++ standard chapter, where it's mentioned?

Reply:
I think this is the relevant part of the standard: https://eel.is/c++draft/basic.life#1.1

TL;DR casting and then dereferencing a pointer is only valid if you guarantee proper object size, layout, and alignment. So (let's forget about inheritance for now), in general you can only safely cast a pointer to something, that has the size of 1 byte (and treat it as byte array), and then back to the same type. Now, the byte size is the tricky part. Although modern systems agree that 1 byte == 8 bits, that hasn't always been the case (an interesting [wiki article](https://en.wikipedia.org/wiki/Byte) on that topic). So, from language point of view, one can not assume that `sizeof(uint8_t) == 1`. But `unsigned char`, `char` and `std::byte` are guaranteed to have size of 1 byte. C++ 20 brings some confusion by declaring `sizeof(char8_t) == 1`, but SYCL 2020 is based on C++ 17. This is how I interpret the standard, I may be wrong about that. Hopefully @Pennycook can correct me.

Reply:
I agree with everything @alexbatashev said.  @capatober, I think the easiest fix is to replace the `uint8_t` in the spec with a `std::byte` -- everything else is fine.  Sorry for not picking up on this during review of the spec; things didn't click until I saw the scratchpad being reinterpreted.

Reply:
@Pennycook and what do you think of embracing `sycl::span` instead of raw pointers? I think, it can help with safety, given that users will be forced to provide size for each range.

Reply:
I'm concerned that the `sycl::span` variant would be harder to use in the case where the same scratchpad is used for multiple sorts (and, eventually, multiple algorithms).

With a raw pointer / span of bytes, the following usage is allowed (where `array1` and `array2` have different types).
```c++
auto exec = group_with_scratchpad(id.get_group(), span(&scratch[0], local_memory_size));
joint_sort(exec, array1, array1 + N);
joint_sort(exec, array2, array2 + N);
```

The implementation can use `placement new` to allocate objects inside of the `scratch` memory and everything is fine.  Additionally, the implementation can allocate anything it wants in that scratchpad, not just objects of the type being sorted.

If the user has to pass a `span` associated with a particular type, wouldn't they have to handle correct setup and teardown of the memory either side of calling `joint_sort`?  Presumably, the implementation would also lose the freedom to use the memory for anything except objects of the type declared in the `span`.

Reply:
Just my opinion: if 8-bits not equal to 1 byte, then many places where SYCL would also not work properly and would never reached the point that is supposed to start some sorting algorithm ))
For example, take a look at this code: https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/aliases.hpp#L85

Reply:
@v-klochkov this is OpenCL C type. It is specified to be 8 bits long, rather than 1 byte. See section 6.3.1 of OpenCL C spec.

Reply:
Quite interesting discussion. I remember words in the C++ standard about guaranteed object size, but I couldn't realize that 8 bits != 1 byte :)